src/atdf/patch.rs

```diff
@@ -62,3 +62,49 @@ pub fn remove_unsafe_cpu_regs(chip: &mut chip::Chip, _el: &xmltree::Element) ->
 
     Ok(())
 }
+
+pub fn remove_register_common_prefix(chip: &mut chip::Chip) -> crate::Result<()> {
+    for peripheral in chip.peripherals.values_mut() {
+        let mut common_prefix = String::new();
+        let mut prefix_index = 0;
+
+        'outer: loop {
+            // There's not enough quorum in less than two elements to find a
+            // prefix.
+            if peripheral.registers.len() < 2 {
+                break 'outer;
+            }
+
+            let mut c = None;
+            for register in peripheral.registers.values() {
+                if register.name.chars().count() == prefix_index {
+                    break 'outer;
+                }
+
+                let current_char = register.name.chars().nth(prefix_index).unwrap();
+
+                match c {
+                    None => c = Some(current_char),
+                    Some(c) if c != current_char => break 'outer,
+                    _ => continue,
+                }
+            }
+
+            if let Some(c) = c {
+                common_prefix.push(c);
+            }
+            prefix_index += 1;
+        }
+
+        let is_valid_prefix = common_prefix.ends_with("_") && common_prefix.chars().count() >= 2;
+        if is_valid_prefix {
+            for register in peripheral.registers.values_mut() {
+                if let Some(s) = register.name.strip_prefix(&common_prefix) {
+                    register.name = s.to_string();
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
```

Comment:
To me this function mixes the "business logic" and prefix-finding-algorithm too much.  It is really hard to understand what is going on.  I would at least pull the code for finding the longest prefix into its own function.  And maybe implement it using less character juggling and more using string functions.  This may be less efficient but will be _much_ easier to follow...

So something like the following (untested)

```rust
// this is just what I could come up with, not sure if there is a more readable
// option
fn longest_common_prefix<'a>(strings: &[&'a str]) -> &'a str {
    if strings.len() == 0 {
        return "";
    }

    let mut longest_prefix = "";
    for prefix in strings[0].char_indices().map(|(i, _)| strings[0].split_at(i).0) {
        if strings.iter().all(|s| s.starts_with(prefix)) {
            longest_prefix = prefix;
        } else {
            // this prefix no longer matches so the previous one was the longest
            break;
        }
    }
    longest_prefix
}

pub fn remove_register_common_prefix(chip: &mut chip::Chip) -> crate::Result<()> {
    for peripheral in chip.peripherals.values_mut() {
        // There's not enough quorum in less than two elements to find a
        // prefix.
        if peripheral.registers.len() < 2 {
            continue;
        }

        let register_names: Vec<_> = peripheral.registers.keys().map(String::as_str).collect();
        let common_prefix = longest_common_prefix(&register_names).to_string();

        if common_prefix.ends_with("_") && common_prefix.chars().count() >= 2 {
            for register in peripheral.registers.values_mut() {
                if let Some(s) = register.name.strip_prefix(&common_prefix) {
                    register.name = s.to_string();
                }
            }
        }
    }

    Ok(())
}
```

Reply:
>To me this function mixes the "business logic" and prefix-finding-algorithm too much. It is really hard to understand what is going on. I would at least pull the code for finding the longest prefix into its own function. And maybe implement it using less character juggling and more using string functions. This may be less efficient but will be much easier to follow...

I do agree that the code is kind of ugly, did it that way to make it faster as each iteration as in some peripherals there are lots of registers and avoid intermediate allocations. (also, your code showed me that allocations can be removed in `common_prefix` 👀 ).

I adapted the function from a being a generic function that took an str array and inlined it to avoid `collect`s.

I'll do it in separate ways to sacrifice performance for readability.

Reply:
Done, tested with your code and has the same behavior. Didn't measure the difference between the two but at least on my machine it doesn't have much of an impact.